crates/bevy_graph/src/graphs/mod.rs

```diff
@@ -0,0 +1,174 @@
+    /// # let mut graph = SimpleMapGraph::<(), i32, true>::new();
+    /// let from = graph.new_node(());
+    /// let to = graph.new_node(());
+    /// let edge = graph.new_edge(from, to, 12).unwrap();
+    /// assert_eq!(edge.get(&graph).unwrap(), &12);
+    /// ```
+    fn get_edge(&self, edge: EdgeIdx) -> Result<&E, GraphError>;
+
+    /// Returns a mutable reference to the value of a given edge
+    ///
+    /// ## Inline helper
+    ///
+    /// This function can also be directly called by the `EdgeIdx`:
+    /// ```
+    /// # use bevy_graph::graphs::{Graph, simple::SimpleMapGraph};
+    /// # let mut graph = SimpleMapGraph::<(), i32, true>::new();
+    /// let from = graph.new_node(());
+    /// let to = graph.new_node(());
+    /// let edge = graph.new_edge(from, to, 12).unwrap();
+    /// assert_eq!(edge.get_mut(&mut graph).unwrap(), &12);
+    /// ```
+    fn get_edge_mut(&mut self, edge: EdgeIdx) -> Result<&mut E, GraphError>;
+
+    /// Remove an edge by its `EdgeIdx` and returns the edge data
+    ///
+    /// ## Inline helper
+    ///
+    /// This function can also be directly called by the `EdgeIdx`:
+    /// ```
+    /// # use bevy_graph::graphs::{Graph, simple::SimpleMapGraph};
+    /// # let mut graph = SimpleMapGraph::<(), i32, true>::new();
+    /// let from = graph.new_node(());
+    /// let to = graph.new_node(());
+    /// let edge = graph.new_edge(from, to, 12).unwrap();
+    /// assert_eq!(edge.remove(&mut graph).unwrap(), 12);
+    /// ```
+    fn remove_edge(&mut self, edge: EdgeIdx) -> Result<E, GraphError>;
+
+    /// Remove an edge by its `EdgeIdx` and returns the edge data
+    ///
+    /// # Safety
+    ///
+    /// This function should only be called when the edge exists
+    unsafe fn remove_edge_unchecked(&mut self, edge: EdgeIdx) -> E;
+
+    /// Returns a `Vec` of all edges between two nodes as `EdgeIdx`
+    fn edges_between(&self, from: NodeIdx, to: NodeIdx) -> Result<Vec<EdgeIdx>, GraphError>;
+
+    /// Returns a `Vec` of all edges between two nodes as `EdgeIdx`
+    ///
+    /// # Safety
+    ///
+    /// This function should only be called when the nodes and the edges between exists
+    unsafe fn edges_between_unchecked(&self, from: NodeIdx, to: NodeIdx) -> Vec<EdgeIdx>;
+
+    /// Returns a `Vec` of all edges the node is outgoing from.
+    fn edges_of(&self, node: NodeIdx) -> Vec<(NodeIdx, EdgeIdx)>; // TODO: can we use other type than Vec? maybe directly iterator?
+
+    ////////////////////////////
+    // Algos
+    ////////////////////////////
+
+    /// Makes a `BreadthFirstSearch` over this graph
+    #[inline]
+    fn breadth_first_search(&self, start: NodeIdx) -> BreadthFirstSearch {
+        BreadthFirstSearch::with_capacity(start, self.count())
+    }
+
+    /// Makes a `DepthFirstSearch` over this graph
+    #[inline]
+    fn depth_first_search(&self, start: NodeIdx) -> DepthFirstSearch {
+        DepthFirstSearch::with_capacity(start, self.count())
+    }
+}
```

Comment:
I would keep algorithms out of these traits.

IMO, algorithms probably don't need to be trait methods. `Bfs::new(&graph)` wouldn't look that bad. Otherwise, you can write another trait to hold the methods and blanket implement it for all types that implement `Graph<N, E>`.